File: frontend/src/stores/evalRunbook.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { api } from "@/services/api";

export interface ScoreBreakdownItem {
  points: number;
  notes: string;
}

export interface FMVComparable {
  title: string;
  price: number;
  condition?: string;
  days_ago?: number;
}

export interface EvalRunbook {
  id: number;
  book_id: number;
  total_score: number;
  score_breakdown: Record<string, ScoreBreakdownItem>;
  recommendation: "PASS" | "ACQUIRE";
  original_asking_price?: number;
  current_asking_price?: number;
  discount_code?: string;
  price_notes?: string;
  fmv_low?: number;
  fmv_high?: number;
  recommended_price?: number;
  ebay_comparables?: FMVComparable[];
  abebooks_comparables?: FMVComparable[];
  condition_grade?: string;
  condition_positives?: string[];
  condition_negatives?: string[];
  critical_issues?: string[];
  analysis_narrative?: string;
  item_identification?: Record<string, string>;
  generated_at: string;
  created_at: string;
  updated_at: string;

  // Tiered recommendation fields
  recommendation_tier?: "STRONG_BUY" | "BUY" | "CONDITIONAL" | "PASS";
  quality_score?: number;
  strategic_fit_score?: number;
  combined_score?: number;
  price_position?: "EXCELLENT" | "GOOD" | "FAIR" | "POOR";
  suggested_offer?: number;
  recommendation_reasoning?: string;
  strategic_floor_applied?: boolean;
  quality_floor_applied?: boolean;
  scoring_version?: string;
  score_source?: "eval_runbook" | "napoleon";
  last_scored_price?: number;

  // Napoleon override fields
  napoleon_recommendation?: string;
  napoleon_reasoning?: string;
  napoleon_analyzed_at?: string;
}

export interface PriceUpdatePayload {
  new_price: number;
  discount_code?: string;
  notes?: string;
}

export interface PriceUpdateResponse {
  previous_price?: number;
  new_price: number;
  score_before: number;
  score_after: number;
  recommendation_before: string;
  recommendation_after: string;
  runbook: EvalRunbook;
}

export interface PriceHistoryEntry {
  id: number;
  previous_price?: number;
  new_price?: number;
  discount_code?: string;
  notes?: string;
  score_before?: number;
  score_after?: number;
  changed_at: string;
}

export interface RefreshResponse {
  status: "completed" | "failed";
  score_before?: number;
  score_after: number;
  message: string;
  runbook: EvalRunbook;
}

export const useEvalRunbookStore = defineStore("evalRunbook", () => {
  const currentRunbook = ref<EvalRunbook | null>(null);
  const priceHistory = ref<PriceHistoryEntry[]>([]);
  const loading = ref(false);
  const error = ref<string | null>(null);

  async function fetchRunbook(bookId: number): Promise<EvalRunbook | null> {
    loading.value = true;
    error.value = null;
    try {
      const response = await api.get(`/books/${bookId}/eval-runbook`);
      currentRunbook.value = response.data;
      return response.data;
    } catch (e: unknown) {
      const err = e as { response?: { status?: number }; message?: string };
      if (err.response?.status === 404) {
        currentRunbook.value = null;
        return null;
      }
      error.value = err.message || "Failed to fetch eval runbook";
      throw e;
    } finally {
      loading.value = false;
    }
  }

  async function updatePrice(
    bookId: number,
    payload: PriceUpdatePayload
  ): Promise<PriceUpdateResponse> {
    loading.value = true;
    error.value = null;
    try {
      const response = await api.patch(`/books/${bookId}/eval-runbook/price`, payload);
      currentRunbook.value = response.data.runbook;
      return response.data;
    } catch (e: unknown) {
      const message = e instanceof Error ? e.message : "Failed to update price";
      error.value = message;
      throw e;
    } finally {
      loading.value = false;
    }
  }

  async function fetchPriceHistory(bookId: number): Promise<PriceHistoryEntry[]> {
    try {
      const response = await api.get(`/books/${bookId}/eval-runbook/history`);
      priceHistory.value = response.data;
      return response.data;
    } catch (e: unknown) {
      console.error("Failed to fetch price history:", e);
      return [];
    }
  }

  async function refreshRunbook(bookId: number): Promise<RefreshResponse> {
    loading.value = true;
    error.value = null;
    try {
      const response = await api.post(`/books/${bookId}/eval-runbook/refresh`);
      currentRunbook.value = response.data.runbook;
      return response.data;
    } catch (e: unknown) {
      const err = e as { response?: { data?: { detail?: string } }; message?: string };
      error.value = err.response?.data?.detail || err.message || "Failed to refresh analysis";
      throw e;
    } finally {
      loading.value = false;
    }
  }

  function clearRunbook() {
    currentRunbook.value = null;
    priceHistory.value = [];
    error.value = null;
  }

  return {
    currentRunbook,
    priceHistory,
    loading,
    error,
    fetchRunbook,
    updatePrice,
    fetchPriceHistory,
    refreshRunbook,
    clearRunbook,
  };
});

File: backend/app/schemas/eval_runbook.py
"""Eval Runbook schemas."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class ScoreBreakdownItem(BaseModel):
    """Individual scoring criterion."""

    points: int
    notes: str


class FMVComparable(BaseModel):
    """A comparable listing for FMV calculation."""

    title: str
    price: Decimal | None = None  # Can be None if extraction failed
    condition: str | None = None
    days_ago: int | None = None  # For eBay sold


class EvalRunbookBase(BaseModel):
    """Base eval runbook schema."""

    total_score: int
    score_breakdown: dict[str, ScoreBreakdownItem]
    recommendation: str = Field(pattern="^(PASS|ACQUIRE)$")

    # Pricing
    original_asking_price: Decimal | None = None
    current_asking_price: Decimal | None = None
    discount_code: str | None = None
    price_notes: str | None = None
    fmv_low: Decimal | None = None
    fmv_high: Decimal | None = None
    fmv_notes: str | None = None
    fmv_confidence: str | None = None  # "high", "medium", "low"
    recommended_price: Decimal | None = None

    # Tiered recommendation fields
    recommendation_tier: str | None = None  # STRONG_BUY, BUY, CONDITIONAL, PASS
    quality_score: int | None = None  # 0-100
    strategic_fit_score: int | None = None  # 0-100
    combined_score: int | None = None  # Weighted average
    price_position: str | None = None  # EXCELLENT, GOOD, FAIR, POOR
    suggested_offer: Decimal | None = None  # For CONDITIONAL recommendations
    recommendation_reasoning: str | None = None  # 1-2 sentence explanation
    strategic_floor_applied: bool = False
    quality_floor_applied: bool = False
    scoring_version: str = "2025-01"
    score_source: str = "eval_runbook"  # "eval_runbook" or "napoleon"
    last_scored_price: Decimal | None = None

    # Napoleon Analysis override fields
    napoleon_recommendation: str | None = None
    napoleon_reasoning: str | None = None
    napoleon_analyzed_at: datetime | None = None

    # Comparables
    ebay_comparables: list[FMVComparable] | None = None
    abebooks_comparables: list[FMVComparable] | None = None

    # Content
    condition_grade: str | None = None
    condition_positives: list[str] | None = None
    condition_negatives: list[str] | None = None
    critical_issues: list[str] | None = None
    analysis_narrative: str | None = None
    item_identification: dict | None = None


class EvalRunbookResponse(EvalRunbookBase):
    """Eval runbook response schema."""

    id: int
    book_id: int
    generated_at: datetime
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EvalRunbookPriceUpdate(BaseModel):
    """Schema for updating eval runbook price."""

    new_price: Decimal
    discount_code: str | None = None
    notes: str | None = None


class EvalRunbookPriceUpdateResponse(BaseModel):
    """Response after price update with recalculated score."""

    previous_price: Decimal | None
    new_price: Decimal
    score_before: int
    score_after: int
    recommendation_before: str
    recommendation_after: str
    runbook: EvalRunbookResponse


class EvalPriceHistoryResponse(BaseModel):
    """Price history entry response."""

    id: int
    previous_price: Decimal | None
    new_price: Decimal | None
    discount_code: str | None
    notes: str | None
    score_before: int | None
    score_after: int | None
    changed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EvalRunbookRefreshResponse(BaseModel):
    """Response after refresh with full AI analysis."""

    status: str  # "completed" or "failed"
    score_before: int | None
    score_after: int
    message: str
    runbook: EvalRunbookResponse
